src/cascadia/PublicTerminalCore/HwndTerminal.cpp

```diff
@@ -0,0 +1,413 @@
+
+    return S_OK;
+}
+
+void HwndTerminal::RegisterScrollCallback(std::function<void(int, int, int)> callback)
+{
+    _terminal->SetScrollPositionChangedCallback(callback);
+}
+
+void HwndTerminal::RegisterWriteCallback(void _stdcall callback(wchar_t*))
+{
+    _terminal->SetWriteInputCallback([=](std::wstring & input) noexcept {
+        const wchar_t* text = input.c_str();
+        const size_t textChars = wcslen(text) + 1;
+        const size_t textBytes = textChars * sizeof(wchar_t);
+        wchar_t* callingText = nullptr;
+
+        callingText = static_cast<wchar_t*>(::CoTaskMemAlloc(textBytes));
+
+        if (callingText == nullptr)
+        {
+            callback(nullptr);
+        }
+        else
+        {
+            wcscpy_s(callingText, textChars, text);
+
+            callback(callingText);
+        }
+    });
+}
+
+void HwndTerminal::_UpdateFont(int newDpi)
+{
+    auto lock = _terminal->LockForWriting();
+
+    // TODO: MSFT:20895307 If the font doesn't exist, this doesn't
+    //      actually fail. We need a way to gracefully fallback.
+    _renderer->TriggerFontChange(newDpi, _desiredFont, _actualFont);
+}
+
+HRESULT HwndTerminal::Refresh(double width, double height, _Out_ int* charColumns, _Out_ int* charRows)
+{
+    RETURN_HR_IF_NULL(E_INVALIDARG, charColumns);
+    RETURN_HR_IF_NULL(E_INVALIDARG, charRows);
+
+    auto lock = _terminal->LockForWriting();
+    const SIZE windowSize{ static_cast<short>(width), static_cast<short>(height) };
+    RETURN_IF_FAILED(_renderEngine->SetWindowSize(windowSize));
+
+    // Invalidate everything
+    _renderer->TriggerRedrawAll();
+
+    // Convert our new dimensions to characters
+    const auto viewInPixels = Viewport::FromDimensions({ 0, 0 },
+                                                       { gsl::narrow<short>(windowSize.cx), gsl::narrow<short>(windowSize.cy) });
+    const auto vp = _renderEngine->GetViewportInCharacters(viewInPixels);
+
+    // If this function succeeds with S_FALSE, then the terminal didn't
+    //      actually change size. No need to notify the connection of this
+    //      no-op.
+    // TODO: MSFT:20642295 Resizing the buffer will corrupt it
+    // I believe we'll need support for CSI 2J, and additionally I think
+    //      we're resetting the viewport to the top
+    RETURN_IF_FAILED(_terminal->UserResize({ vp.Width(), vp.Height() }));
+    *charColumns = vp.Width();
+    *charRows = vp.Height();
+
+    return S_OK;
+}
+
+void HwndTerminal::SendOutput(std::wstring_view data)
+{
+    _terminal->Write(data);
+}
+
+HRESULT _stdcall CreateTerminal(HWND parentHwnd, _Out_ void** hwnd, _Out_ void** terminal)
+{
+    auto _terminal = std::make_unique<HwndTerminal>(parentHwnd);
+    RETURN_IF_FAILED(_terminal->Initialize());
+
+    *hwnd = _terminal->_hwnd.get();
+    *terminal = _terminal.release();
+
+    return S_OK;
+}
+
+void _stdcall RegisterScrollCallback(void* terminal, void __stdcall callback(int, int, int))

```

Comment:
I don't want to block for this specifically, but I feel like these APIs should be namespaced (something as simple as `TerminalRegisterScrollCallback` and friends). Otherwise they're just loud bare _C linkage_ exports that don't explain anything through their names.